### PR TITLE
Possible naive fix for issue #321

### PR DESCRIPTION
**Pull request by labriola**
**What kind of change does this PR introduce?**
Potentially naive fix for issue #321 

**Did you add tests for your changes?**
No tests added as the issue only manifests under HMR and didn't see a correlated way to add those in the transform folder.

**If relevant, did you update the README?**
I don't believe it's relevant at this time

**Summary**
Ensures that transform function, when defined, is applied to both initial load and updates via HMR

Potentially naive fix for issue #321 

**Does this PR introduce a breaking change?**

The only possibility is if someone was expecting an initial load and subsequent reload of a CSS file to be different. I don't know of a use case.

### CODE REVIEW

**Comment by labriola** [comment]
This branch now has tests.... Unfortunately without major refactoring there is no way to write the tests to the degree I would like. There is some package level memoization and caching that make things more stateful than they should be across tests. Also seeing a new issue, filed as #324 that was not possible to fix at this time.

File: test/addStyles.test.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      jsdomHtml,
+      undefined,
+      0,
+      done,
+      applyUpdate
+    )
+   });
+
+  // See issue #324, place holder for a future test once that issue is fixed
+  it.skip("should load the css if the transform function returns data during HMR");
```
Comment (evilebottnawi) [inline test/addStyles.test.js]: What actually we need to solve this problem?

Comment (labriola) [reply test/addStyles.test.js]: The discussion we had over on issue #324 there is quite a bit of refactoring. I can remove the test if you want but I am not in a position to spend more time on this at present

File: lib/addStyles.js [hunk 2]
```diff
@@ -107,7 +107,6 @@ module.exports = function(list, options) {
 
 			if(domStyle.refs === 0) {
 				for (var j = 0; j < domStyle.parts.length; j++) domStyle.parts[j]();
-
 				delete stylesInDom[domStyle.id];
```
[the remark this answers is not in the document]
Comment (labriola) [reply lib/addStyles.js]: It was unintentional, I had a debugger statement there and, when I removed it, the newline was removed too. Sorry. 

The tests I have right now cover the changes but were extremely difficult to get in place with this module. If you are interested I can try to help refactor addstyles in the future, but I can't imagine adding more tests to this PR at the moment

File: test/addStyles.test.js (hunk 1 above)
Comment (evilebottnawi) [reply test/addStyles.test.js]: @labriola you can left this test, it is normal, we should add more tests to increase coverage

**Comment by labriola** [comment]
Requests for fix to line break and tab resolved.

**Comment by labriola** [comment]
I also just fixed a mistake which was causing the code coverage to drop so much.

**Comment by evilebottnawi** [comment]
@labriola hm strange `(-26.16%)` :confused:

**Comment by evilebottnawi** [comment]
In short: now transform works always as expected, exclude for hmr right?

**Comment by labriola** [comment]
@evilebottnawi the code coverage only appeared to drop as there were never any tests for addStyles before. Now that there are, its counting addStyle as part of the total. So, we have increased test coverage but decreased the reported percentage.

In short, transform works all the time now, including during hmr.

There is one edge case left during hmr . If the first time the module is processed the transform returns falsey, the module will never be added. This is an edge case though and probably not likely, I just like to be thorough so I documented it as  issues #324

**Comment by labriola** [comment]
I added the test for issue #324 as requested. It is set to skip() now but I will try to address it in a separate PR when possible. For now, it will demonstrate the issue.

**Comment by evilebottnawi** [comment]
@labriola I have already requested an additional review from other contributors (just wait, i also don't love long merges, but it is package uses very very very many people and it is require more times on review ), do not worry. The only thing I'm afraid is that they can reject because due hmr

**Comment by mauricecruz** [comment]
Resurfacing this PR. Running across the same issue and hoping I don't need to fork either.

Please let me know if there's anything I can do to move this along.

File: test/addStyles.test.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      jsdomHtml,
+      undefined,
+      0,
+      done,
+      applyUpdate
+    )
+   });
+
+  // See issue #324, place holder for a future test once that issue is fixed
+  it.skip("should load css if the transform function intially returns false but returns data during subsequent HMR", function(done) {
```
Comment (evilebottnawi) [inline test/addStyles.test.js]: We should fix this problem here

**Comment by labriola** [comment]
//cc @evilebottnawi Quick reminder. This PR is _not_ waiting on me. 

I increased the overall test coverage as my PR covers my code completely. You are seeing a decline in the project test coverage _percentage_ as my PR _exposes_ the fact that this project is completely missing test coverage of critical modules. Please see comment thread ending on May 4, 2018

If you would like, I will happily resubmit my PR, and remove all of my test. The net result will be that your test coverage will go up dramatically. However, I am not adding test coverage to unrelated code that I did not modify in this particular PR. It makes the purpose of this PR unclear and I would consider it an anti-pattern.

**Comment by evilebottnawi** [comment]
@labriola thanks for clarify, i will review this in near future

**Comment by evilebottnawi** [comment]
Sorry for delay, and sorry we can't merge this because transform was removed without replacement

**Comment by labriola** [comment]
Amusing end to a 1.25 year saga to get a PR merged. At least we have reached a resolution

**Comment by evilebottnawi** [comment]
@labriola it is open source and sometimes do not have enough time on all PRs/issues, sorry, we prepare style-loader/css-loader to built-in in webpack, so many option was removed

**Comment by labriola** [comment]
@evilebottnawi I really do understand but that is a self-perpetuating problem. Had we been able to get this PR merged, I would have fixed a dozen more issues and been happy to help more. Unfortunately, if projects discourage (actively or passively) people from contributing, they end up with less and less time as projects grow. I am not criticizing you personally and I really do empathize with time constraints, I just see it as unfortunate. That said, best of luck, I will look at what comes next in webpack and try to fix any issues there.
